src/evict.c

```diff
@@ -398,11 +398,12 @@ int getMaxmemoryState(size_t *total, size_t *logical, size_t *tofree, float *lev
     if (total) *total = mem_reported;
 
     /* We may return ASAP if there is no need to compute the level. */
-    if (!server.maxmemory) {
+    if (!server.maxmemory_available) {
```

Comment:
Thinking about this more, I feel that it is more preferable to model this new setting as a "soft" `maxmemory`, which can trigger key eviction earlier but won't cause OOM to be returned before hitting the actual `maxmemory`. Otherwise, we effectively create an alias of `maxmemory`. More specifically, I think `performEviction` should only return `EVICT_FAIL` when the memory usage goes beyond the real `maxmemory`.

Additionally, since `getMaxmemoryState` is also used outside of the OOM prevention path such as in `VM_GetUsedMemoryRatio`, we should consider parameterizing `maxmemory` and having it passed in by the caller instead, so that we can maintain the same semantics in these externally facing scenarios. 

Thoughts?

Reply:
Like what you suggested, if we can think "maxmemory_available" is available, we can think this is a soft maxmemory. However you maybe think we should not return OOM, then in the case, how we can return message to client, any idea?

Reply:
right - how about just processing the command normally?

1. if `used_memory` is below soft max, no change to the existing logic
2. if `used_memory` is above soft max but below hard max, trigger key eviction and continue with the normal processing
3. if `used_memory` is above hard max, no change to the existing logic, i.e., trigger key eviction and fail the command if the `used_memory` is still above hard max after the eviction)